src/proposer.ts

```diff
@@ -121,6 +130,246 @@ let s: ReturnType<typeof strings>
 // Initialization flag
 let isInitialized = false
 
+// ~7 days at ~12s blocks
+const APPROX_7D_BLOCKS = 50400
+/**
+ * Validates that a vault ID exists on-chain by checking it is within range
+ * [1, nextVaultId - 1]. Throws if invalid or out of range.
+ */
+export async function validateVaultIdOnChain(vaultId: number): Promise<void> {
+	// Basic sanity
+	if (!Number.isInteger(vaultId) || vaultId < 1) {
+		throw new Error('Invalid vault ID')
+	}
+	// Ensure contracts are initialized
+	if (!vaultTrackerContract) {
+		throw new Error('VaultTracker contract not initialized')
+	}
+	const nextId = await vaultTrackerContract.nextVaultId()
+	const nextIdNumber = Number(nextId)
+	if (!Number.isFinite(nextIdNumber)) {
+		throw new Error('Could not determine nextVaultId from chain')
+	}
+	if (vaultId >= nextIdNumber) {
+		throw new Error('Vault ID does not exist on-chain')
+	}
+}
+
+/**
+ * Computes block range hex strings for Alchemy getAssetTransfers requests,
+ * defaulting to a ~7 day lookback if not provided.
+ */
+async function computeBlockRange(
+	fromBlockHex?: string,
+	toBlockHex?: string
+): Promise<{ fromBlockHex: string; toBlockHex: string }> {
+	const provider = (await sepoliaAlchemy.config.getProvider()) as unknown as {
+		getBlockNumber: () => Promise<number>
+	}
+	const latest = await provider.getBlockNumber()
+	const resolvedTo = toBlockHex ?? '0x' + latest.toString(16)
+	const fromBlock = Math.max(0, latest - APPROX_7D_BLOCKS)
+	const resolvedFrom = fromBlockHex ?? '0x' + fromBlock.toString(16)
+	return { fromBlockHex: resolvedFrom, toBlockHex: resolvedTo }
+}
+
+/**
+ * Generic discovery for deposits into VaultTracker using Alchemy's decoded
+ * asset transfers. Supports ERC-20 and ETH (internal/external) categories.
+ */
+
+async function discoverAndIngestDeposits(params: {
+	controller: string
```

Comment:
Does this function really need a `controller`? It might make more sense to sweep up all deposits from any address within a given time period, and discover the `controller` address as we go?

Reply:
Looks like the `fromAddress` is optional, so we could skip it, I think. https://www.alchemy.com/docs/data/transfers-api/transfers-endpoints/alchemy-get-asset-transfers#request.body.assetTransferParams.fromAddress